modules/storages/app/services/storages/group_folder_properties_sync_service.rb

```diff
@@ -28,266 +28,268 @@
-    #   when the project is public and non member has at least one storage permission
-    #   then all non memebers should have access to the project folder
-    if !(project.public? && ProjectRole.non_member.permissions.intersect?(PERMISSIONS_KEYS))
-      tokens_query = tokens_query.where(users: project.users)
+    def ensure_folders_exist(remote_folders)
+      id_folder_map = remote_folders.to_h { |folder, properties| [properties['fileid'], folder] }
+
+      active_project_storages_scope.map do |project_storage|
+        next create_folder(project_storage) unless id_folder_map.key?(project_storage.project_folder_id)
+
+        current_path = id_folder_map[project_storage.project_folder_id]
+        if current_path == project_storage.project_folder_path
+          project_storage.project_folder_id
+        else
+          move_folder(project_storage, current_path)
+            .result_or do |error|
+            error_msg = { command: 'nextcloud.rename_file',
+                          source: current_path,
+                          target: project_storage.project_folder_path,
+                          data: { status: error.data.code, body: error.data.body } }.to_json
+
+            # we need to stop as this would mess with the other processes
+            return OpenProject.logger.warn error_msg
+          end
+        end
+      end
     end
 
-    tokens_query
-  end
+    def move_folder(project_storage, current_name)
+      Peripherals::Registry
+        .resolve("commands.nextcloud.rename_file")
+        .call(storage: @storage, source: current_name, target: project_storage.project_folder_path)
+    end
 
-  def add_active_users_to_group
-    @nextcloud_usernames_used_in_openproject.each do |nextcloud_username|
-      if group_users.exclude?(nextcloud_username)
-        query_params = { user: nextcloud_username }
-        Storages::Peripherals::Registry.resolve("commands.#{@storage.short_provider_type}.add_user_to_group")
-          .call(storage: @storage, **query_params)
-          .on_failure(&failure_handler('add_user_to_group_command', query_params))
+    def create_folder(project_storage)
+      folder_path = project_storage.project_folder_path
+      Peripherals::Registry
+        .resolve("commands.nextcloud.create_folder")
+        .call(storage: @storage, folder_path:)
+        .result_or do |error|
+        error_msg = { command: 'nextcloud.create_folder',
+                      folder: folder_path,
+                      data: { status: error.data.code, body: error.data.body } }.to_json
+
+        return OpenProject.logger.warn(error_msg)
       end
+
+      folder_ids = Peripherals::Registry
+                     .resolve('queries.nextcloud.file_ids')
+                     .call(storage: @storage, path: folder_path)
+                     .result_or { |_error| puts :error } # Log this
+
+      project_storage.update(project_folder_id: folder_ids.dig(folder_path, 'fileid'))
```

Comment:
`folder_ids.dig(folder_path, 'fileid')` can return `nil` potentially and then we are left with nil project_folder_id. Do you think it is a problem? I am not sure, but maybe we should handle the nil case, because later it could be really not easy to debug if it happens.

Reply:
If we update it with nil it won't be used later on so no harm done.